File: backend/app/core/geometry.py
from __future__ import annotations

import math
import numpy as np

from app.core.constants import EARTH_OMEGA, EARTH_RADIUS_KM, EARTH_MU


def finite(x: object) -> bool:
    """Проверка, является ли значение конечным числом."""
    return isinstance(x, (int, float)) and (not isinstance(x, bool)) and math.isfinite(x)


def compute_positions(s: dict, t_s: float) -> tuple[list[str], np.ndarray, np.ndarray]:
    """
    Вычисление координат спутников на момент времени t_s.

    Возвращает кортеж:
    - список идентификаторов спутников;
    - массив координат в инерциальной системе ECI [км];
    - массив координат во вращающейся геоцентрической системе ECEF [км].
    """
    e, d = (s["environment"], s["design"])
    pmap = {p["id"]: p for p in d["planes"]}
    r = EARTH_RADIUS_KM + float(e["altitude_km"])
    # Среднее движение на круговой орбите радиуса r
    n = math.sqrt(EARTH_MU / r**3)
    inc = math.radians(float(e["inclination_deg"]))

    # Аргумент широты спутников: фазирование внутри плоскости и движение по орбите за время t_s
    u = np.array(
        [
            math.radians(float(x["slot_deg"]) + float(pmap[x["plane_id"]]["phase_deg"])) + n * t_s
            for x in d["satellites"]
        ],
        dtype=np.float64,
    )
    om = np.array(
        [math.radians(float(pmap[x["plane_id"]]["raan_deg"])) for x in d["satellites"]],
        dtype=np.float64,
    )

    cu, su, co, so = np.cos(u), np.sin(u), np.cos(om), np.sin(om)
    # Координаты спутников в инерциальной системе ECI
    xyz = r * np.stack(
        (
            co * cu - so * su * math.cos(inc),
            so * cu + co * su * math.cos(inc),
            su * math.sin(inc),
        ),
        axis=1,
    )

    # Вращение Земли вокруг оси Z: переход от инерциальной ECI к геоцентрической ECEF
    th = math.radians(float(e["earth_angle0_deg"])) + EARTH_OMEGA * t_s
    c, ss = math.cos(th), math.sin(th)
    rot = np.array([[c, -ss, 0.0], [ss, c, 0.0], [0.0, 0.0, 1.0]], dtype=np.float64)
    fixed = xyz @ rot
    sat_ids = [x["id"] for x in d["satellites"]]
    return sat_ids, xyz, fixed


def ground_position(g: dict) -> np.ndarray:
    """Вычисление декартовых координат наземного пункта в системе ECEF [км]."""
    lat, lon = math.radians(float(g["lat_deg"])), math.radians(float(g["lon_deg"]))
    return EARTH_RADIUS_KM * np.array(
        [math.cos(lat) * math.cos(lon), math.cos(lat) * math.sin(lon), math.sin(lat)],
        dtype=np.float64,
    )


def ecef_to_geodetic(xyz: np.ndarray) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """
    Преобразование координат ECEF [км] в геодезические величины.

    Возвращает кортеж:
    - широта в градусах;
    - долгота в градусах;
    - высота над поверхностью Земли в километрах.
    """
    x, y, z = xyz[:, 0], xyz[:, 1], xyz[:, 2]
    hypot_xy = np.hypot(x, y)
    lat_deg = np.degrees(np.arctan2(z, hypot_xy))
    lon_deg = np.degrees(np.arctan2(y, x))
    dist_origin = np.hypot(hypot_xy, z)
    alt_km = dist_origin - EARTH_RADIUS_KM
    return lat_deg, lon_deg, alt_km


def snapshot(s: dict, t_s: float, fast_edges_only: bool = False) -> dict:
    """
    Расчет состояния группировки в момент времени t_s.

    Определяет:
    - пространственное положение и признак активности спутников;
    - ребра межспутниковых линий связи с проверкой дальности и затенения Землей;
    - видимость и углы места относительно наземных станций;
    - ребра связи между наземными пунктами и спутниками.
    """
    e, d = (s["environment"], s["design"])
    ids, inertial, xyz = compute_positions(s, t_s)

    failed = {f["satellite_id"] for f in s.get("failures", []) if f["start_s"] <= t_s < f["end_s"]}
    active = np.array(
        [
            sat["launch_batch"] <= d["launch_stage"] and sat["id"] not in failed
            for sat in d["satellites"]
        ],
        dtype=bool,
    )

    # Межспутниковые линии связи
    n_sats = len(ids)
    edges: list[list[str | float]] = []
    if n_sats > 1:
        i, j = np.triu_indices(n_sats, 1)
        both_active = active[i] & active[j]
        i_cand = i[both_active]
        j_cand = j[both_active]
        if len(i_cand) > 0:
            delta = xyz[j_cand] - xyz[i_cand]
            denom = np.sum(delta * delta, axis=1)
            isl_range = float(e["isl_range_km"])
            isl_range_sq = isl_range * isl_range
            range_ok = denom < isl_range_sq
            if np.any(range_ok):
                i_sub = i_cand[range_ok]
                j_sub = j_cand[range_ok]
                delta_sub = delta[range_ok]
                denom_sub = denom[range_ok]
                dist_sub = np.sqrt(denom_sub)
                # Проекция центра Земли на отрезок между спутниками для проверки затенения
                lam = np.clip(
                    -np.sum(xyz[i_sub] * delta_sub, axis=1) / np.maximum(denom_sub, 1e-12),
                    0.0,
                    1.0,
                )
                closest = np.linalg.norm(xyz[i_sub] + lam[:, None] * delta_sub, axis=1)
                ok = (dist_sub < isl_range) & (closest > EARTH_RADIUS_KM)
                for a, b, dd in zip(i_sub[ok], j_sub[ok], dist_sub[ok]):
                    edges.append([ids[a], ids[b], float(dd)])

    # Линии связи с наземными станциями
    elevations: dict[str, dict[str, float]] = {}
    active_indices = np.where(active)[0]
    min_elev = float(e["min_elevation_deg"])

    for g in s.get("ground_sites", []):
        gid = g["id"]
        gp = ground_position(g)
        dif = xyz - gp
        dl = np.linalg.norm(dif, axis=1)
        # Ограничение скалярного произведения для численной устойчивости
        dot = np.clip(dif @ (gp / EARTH_RADIUS_KM) / dl, -1.0, 1.0)
        el = np.degrees(np.arcsin(dot))

        elevations[gid] = {ids[k]: float(el[k]) for k in active_indices}

        is_gw = g.get("role") == "gateway"
        offline = False
        if is_gw:
            offline = any(
                f["gateway_id"] == gid and f["start_s"] <= t_s < f["end_s"]
                for f in s.get("gateway_outages", [])
            )

        vis = (el >= min_elev) & active & (not offline)
        for k in np.where(vis)[0]:
            edges.append([gid, ids[k], float(dl[k])])

    if fast_edges_only:
        return {
            "t_s": t_s,
            "satellites": [],
            "edges": edges,
            "elevation_deg": elevations,
        }

    lat_arr, lon_arr, alt_arr = ecef_to_geodetic(xyz)

    satellites_out = []
    for k, sid in enumerate(ids):
        sat_meta = d["satellites"][k]
        satellites_out.append(
            {
                "id": sid,
                "plane_id": sat_meta.get("plane_id"),
                "launch_batch": sat_meta.get("launch_batch"),
                "x_km": float(xyz[k, 0]),
                "y_km": float(xyz[k, 1]),
                "z_km": float(xyz[k, 2]),
                "lat_deg": float(lat_arr[k]),
                "lon_deg": float(lon_arr[k]),
                "alt_km": float(alt_arr[k]),
                "active": bool(active[k]),
                "failed": bool(sid in failed),
            }
        )

    return {
        "t_s": t_s,
        "satellites": satellites_out,
        "edges": edges,
        "elevation_deg": elevations,
    }

File: backend/crates/cosmo-core/src/geometry.rs
use std::collections::{HashMap, HashSet};

use crate::constants::{EARTH_MU, EARTH_OMEGA, EARTH_RADIUS_KM};
use crate::models::{GroundSite, Plane, SatellitePosition, Scenario, SnapshotResult};

pub fn finite(x: f64) -> bool {
    x.is_finite()
}

/// Вычисление координат спутников на момент времени t_s.
/// Возвращает кортеж:
/// - вектор идентификаторов спутников;
/// - вектор координат ECI [км];
/// - вектор координат ECEF [км].
pub fn compute_positions(
    s: &Scenario,
    t_s: f64,
) -> (Vec<String>, Vec<[f64; 3]>, Vec<[f64; 3]>) {
    let e = &s.environment;
    let d = &s.design;

    let pmap: HashMap<&str, &Plane> = d.planes.iter().map(|p| (p.id.as_str(), p)).collect();

    let r = EARTH_RADIUS_KM + e.altitude_km;
    let n = (EARTH_MU / (r * r * r)).sqrt();
    let inc = e.inclination_deg.to_radians();

    let n_sats = d.satellites.len();
    let mut sat_ids = Vec::with_capacity(n_sats);
    let mut eci = Vec::with_capacity(n_sats);

    for sat in &d.satellites {
        sat_ids.push(sat.id.clone());
        let dummy_plane = Plane {
            id: sat.plane_id.clone(),
            raan_deg: 0.0,
            phase_deg: 0.0,
        };
        let plane = pmap.get(sat.plane_id.as_str()).copied().unwrap_or(&dummy_plane);
        let u = (sat.slot_deg + plane.phase_deg).to_radians() + n * t_s;
        let om = plane.raan_deg.to_radians();

        let cu = u.cos();
        let su = u.sin();
        let co = om.cos();
        let so = om.sin();

        let x = r * (co * cu - so * su * inc.cos());
        let y = r * (so * cu + co * su * inc.cos());
        let z = r * (su * inc.sin());

        eci.push([x, y, z]);
    }

    let th = e.earth_angle0_deg.to_radians() + EARTH_OMEGA * t_s;
    let c = th.cos();
    let ss = th.sin();

    let ecef = eci
        .iter()
        .map(|&[x, y, z]| [x * c + y * ss, -x * ss + y * c, z])
        .collect();

    (sat_ids, eci, ecef)
}

/// Вычисление декартовых координат наземного пункта в системе ECEF [км].
pub fn ground_position(g: &GroundSite) -> [f64; 3] {
    let lat = g.lat_deg.to_radians();
    let lon = g.lon_deg.to_radians();
    [
        EARTH_RADIUS_KM * lat.cos() * lon.cos(),
        EARTH_RADIUS_KM * lat.cos() * lon.sin(),
        EARTH_RADIUS_KM * lat.sin(),
    ]
}

/// Преобразование координат ECEF [км] в широту, долготу (град) и высоту (км).
pub fn ecef_to_geodetic(xyz: &[f64; 3]) -> (f64, f64, f64) {
    let x = xyz[0];
    let y = xyz[1];
    let z = xyz[2];
    let hypot_xy = x.hypot(y);
    let lat_deg = z.atan2(hypot_xy).to_degrees();
    let lon_deg = y.atan2(x).to_degrees();
    let alt_km = hypot_xy.hypot(z) - EARTH_RADIUS_KM;
    (lat_deg, lon_deg, alt_km)
}

/// Расчет состояния группировки в момент времени t_s.
pub fn snapshot(s: &Scenario, t_s: f64, fast_edges_only: bool) -> SnapshotResult {
    let (ids, _inertial, ecef) = compute_positions(s, t_s);
    let n_sats = ids.len();

    let failed_set: HashSet<&str> = s
        .failures
        .iter()
        .filter(|f| f.start_s <= t_s && t_s < f.end_s)
        .map(|f| f.satellite_id.as_str())
        .collect();

    let active: Vec<bool> = s
        .design
        .satellites
        .iter()
        .map(|sat| sat.launch_batch <= s.design.launch_stage && !failed_set.contains(sat.id.as_str()))
        .collect();

    let mut edges: Vec<(String, String, f64)> = Vec::new();

    if n_sats > 1 {
        let isl_range = s.environment.isl_range_km;
        let isl_range_sq = isl_range * isl_range;

        for i in 0..n_sats {
            if !active[i] {
                continue;
            }
            for j in (i + 1)..n_sats {
                if !active[j] {
                    continue;
                }

                let dx = ecef[j][0] - ecef[i][0];
                let dy = ecef[j][1] - ecef[i][1];
                let dz = ecef[j][2] - ecef[i][2];
                let denom = dx * dx + dy * dy + dz * dz;

                if denom < isl_range_sq {
                    let dist = denom.sqrt();
                    let dot_i_delta = ecef[i][0] * dx + ecef[i][1] * dy + ecef[i][2] * dz;
                    let lam = (-dot_i_delta / denom.max(1e-12)).clamp(0.0, 1.0);

                    let cx = ecef[i][0] + lam * dx;
                    let cy = ecef[i][1] + lam * dy;
                    let cz = ecef[i][2] + lam * dz;
                    let closest = (cx * cx + cy * cy + cz * cz).sqrt();

                    if dist < isl_range && closest > EARTH_RADIUS_KM {
                        edges.push((ids[i].clone(), ids[j].clone(), dist));
                    }
                }
            }
        }
    }

    let mut elevations: HashMap<String, HashMap<String, f64>> = HashMap::new();
    let min_elev = s.environment.min_elevation_deg;

    for g in &s.ground_sites {
        let gid = &g.id;
        let gp = ground_position(g);

        let is_gw = g.role == "gateway";
        let offline = is_gw
            && s.gateway_outages
                .iter()
                .any(|f| f.gateway_id == *gid && f.start_s <= t_s && t_s < f.end_s);

        let mut site_elevs = HashMap::new();

        for k in 0..n_sats {
            if !active[k] {
                continue;
            }

            let dif = [
                ecef[k][0] - gp[0],
                ecef[k][1] - gp[1],
                ecef[k][2] - gp[2],
            ];
            let dl = (dif[0] * dif[0] + dif[1] * dif[1] + dif[2] * dif[2]).sqrt();
            let dot_val = ((dif[0] * gp[0] + dif[1] * gp[1] + dif[2] * gp[2]) / EARTH_RADIUS_KM) / dl;
            let dot_clamped = dot_val.clamp(-1.0, 1.0);
            let el = dot_clamped.asin().to_degrees();

            site_elevs.insert(ids[k].clone(), el);

            if el >= min_elev && !offline {
                edges.push((gid.clone(), ids[k].clone(), dl));
            }
        }

        elevations.insert(gid.clone(), site_elevs);
    }

    if fast_edges_only {
        return SnapshotResult {
            t_s,
            satellites: vec![],
            edges,
            elevation_deg: elevations,
        };
    }

    let mut satellites_out = Vec::with_capacity(n_sats);
    for k in 0..n_sats {
        let sid = &ids[k];
        let sat_meta = &s.design.satellites[k];
        let (lat, lon, alt) = ecef_to_geodetic(&ecef[k]);

        satellites_out.push(SatellitePosition {
            id: sid.clone(),
            plane_id: Some(sat_meta.plane_id.clone()),
            launch_batch: Some(sat_meta.launch_batch),
            x_km: ecef[k][0],
            y_km: ecef[k][1],
            z_km: ecef[k][2],
            lat_deg: lat,
            lon_deg: lon,
            alt_km: alt,
            active: active[k],
            failed: failed_set.contains(sid.as_str()),
        });
    }

    SnapshotResult {
        t_s,
        satellites: satellites_out,
        edges,
        elevation_deg: elevations,
    }
}
